File: src/Movie.ts
import * as Chunk from "./chunk";
import {InvalidDirectorFileError, PathTooNewError} from "./errors";
import {DataStream} from "./DataStream";
import {Cast} from "./Cast";

/* Movie */

export class Movie {
    chunkArrays: Map<string, Chunk.Chunk[]>;
    chunkMap: Chunk.Chunk[];
    _chunkBuffers: Map<Chunk.Chunk, Buffer>;

    castMap: Cast[];
    castMap1: Cast[];
    castMap2: Cast[];

    read(buffer: Buffer) {
        this.chunkArrays = new Map();
        this.chunkMap = [];
        this._chunkBuffers = new Map();
        this.castMap = [];
        this.castMap1 = [];
        this.castMap2 = [];

        var dataStream = new DataStream(buffer);
        dataStream.endianness = DataStream.BIG_ENDIAN; // we set this properly when we create the RIFX chunk
        this.lookupMmap(dataStream);
        this.createCasts();
        this.linkScripts();
    }

    lookupMmap(dataStream: DataStream) {
        // at the beginning of the file, we need to break some of the typical rules. We don't know names, lengths and offsets yet.

        // valid length is undefined because we have not yet reached mmap
        // however, it will be filled automatically in chunk's constructor
        let meta: Chunk.Meta = this.readChunk(dataStream, "RIFX");
        // we can only open DIR or DXR
        // we'll read Movie from dataStream because Movie is an exception to the normal rules
        if (meta.codec != "MV93") {
            throw new PathTooNewError("Codec " + meta.codec + " unsupported.");
        }

        // the next chunk should be imap
        // this HAS to be dataStream for the OFFSET check to be correct
        // we will continue to use it because in this implementation RIFX doesn't contain it
        let imap: Chunk.InitialMap = this.readChunk(dataStream, "imap", null, 12);

        // sanitize mmaps
        /*
        this.differenceImap = 0;
        if (imap.memoryMapArray[0] - 0x2C) {
            this.differenceImap = imap.memoryMapArray[0] - 0x2C;
            for (let i = 0, l = imap.memoryMapArray.length; i < l; i++) {
                imap.memoryMapArray[i] -= this.differenceImap;
            }
        }
        */

        for (let memoryMapOffset of imap.memoryMapArray) {
            dataStream.seek(memoryMapOffset);
            let mmap: Chunk.MemoryMap = this.readChunk(dataStream, "mmap", null, memoryMapOffset);
            // add chunks in the mmap to the chunkArrays HERE
            // make sure to account for chunks with existing names, lengths and offsets
            for (let i = 0, l = mmap.mapArray.length; i < l; i++) {
                let mapEntry = mmap.mapArray[i];
                let chunk;
                if (mapEntry.name === "RIFX") {
                    chunk = meta;
                } else if (mapEntry.name === "imap") {
                    chunk = imap;
                } else if (mapEntry.name === "mmap") {
                    chunk = mmap;
                } else {
                    dataStream.seek(mapEntry.offset);
                    chunk = this.readChunk(dataStream, mapEntry.name, mapEntry.len, mapEntry.offset, mapEntry.padding, mapEntry.unknown0, mapEntry.link);
                }
                this.addChunk(mapEntry.name, chunk);
                this.chunkMap[mapEntry.index] = chunk;
            }
        }
    }

    addChunk(name: string, chunk: Chunk.Chunk) {
        let chunkArray: Chunk.Chunk[] = this.chunkArrays.get(name);
        if (!chunkArray) {
            chunkArray = [];
            this.chunkArrays.set(name, chunkArray);
        }
        chunkArray.push(chunk);
    }

    createCasts() {
        let castList = this.chunkArrays.get("MCsL")[0] as Chunk.CastList;
        let castKey = this.chunkArrays.get("KEY*")[0] as Chunk.CastKey;
        for (let entry of castList.entries) {
            let cast = new Cast();
            cast.readDataEntry(entry);
            cast.readKey(castKey, this.chunkMap);
            this.castMap[cast.id] = cast;
            this.castMap1[cast.id1] = cast;
            this.castMap2[cast.id2] = cast;
        }
    }

    linkScripts() {
        let ctxArray = this.chunkArrays.get("LctX") as Chunk.ScriptContext[];
        if (ctxArray) {
            for (let context of ctxArray) {
                let scriptNames = this.chunkMap[context.lnamSectionID] as Chunk.ScriptNames;
                context.scriptNames = scriptNames;
                for (let section of context.sectionMap) {
                    if (section.sectionID > -1) {
                        let script = this.chunkMap[section.sectionID] as Chunk.Script;
                        script.context = context;
                        script.readNames();
                        script.translate();
                        context.scripts.push(script);
                    }
                }
            }
        }
    }

    readChunk(mainDataStream: any, name: string, len: number = null, offset: number = null, padding: number = null, unknown0: number = null, link: number = null) {
        // the offset is checked against, well, our offset
        let validOffset = mainDataStream.position;
        // check if this is the chunk we are expecting
        let validName = mainDataStream.readFourCC();
        if (name == "RIFX") {
            //if (validName.substring(0, 2) == "MZ") {
                // handle Projector HERE
            //}
            if (validName == "XFIR") {
                mainDataStream.endianness = DataStream.LITTLE_ENDIAN;
                validName = "RIFX";
            }
        }
        // check if it has the length the mmap table specifies
        let validLen = mainDataStream.readUint32();

        // use the valid values if mmap hasn't been read yet
        if (name == null) {
            name = validName;
        }
        if (len == null) {
            len = validLen;
        }
        if (offset == null) {
            offset = validOffset;
        }

        // padding can't be checked, so let's give it a default value if we don't yet know it
        if (padding == null) {
            // padding is usually zero
            if (name !== "free" && name !== "junk") {
                padding = 0;
            } else {
                padding = 12;
            }
        }

        // validate chunk
        if (name !== validName || len !== validLen || offset !== validOffset) {
            throw new InvalidDirectorFileError("At offset " + validOffset + ", expected '" + name + "' chunk with a length of " + len + " and offset of " + offset + " but found a '" + validName + "' chunk with a length of " + validLen + ".");
        }

        if (name === "RIFX") {
            // we're going to pretend RIFX is only 12 bytes long
            // this is because offsets are relative to the beginning of the file
            // whereas everywhere else they're relative to chunk start
            len = 4;
        }

        // copy the contents of the chunk to a new DataStream (minus name/length as that's not what offsets are usually relative to)
        let chunkBuffer = mainDataStream.readBytes(len);
        let chunkDataStream = new DataStream(chunkBuffer, mainDataStream.endianness);

        const chunks = {
            "RIFX": Chunk.Meta,
            "imap": Chunk.InitialMap,
            "mmap": Chunk.MemoryMap,
            "LctX": Chunk.ScriptContext,
            "Lnam": Chunk.ScriptNames,
            "Lscr": Chunk.Script,
            "MCsL": Chunk.CastList,
            "KEY*": Chunk.CastKey,
            "CAS*": Chunk.CastAssociations,
            "CASt": Chunk.CastMember
        };

        let result = chunks[name] ? new chunks[name]() : new Chunk.Unimplemented(name);
        result.read(chunkDataStream);

        this._chunkBuffers.set(result, chunkBuffer);

        return result;
    }
}

File: src/movie.cpp
#include <iostream>
#include <stdexcept>

#include "chunk.h"
#include "lingo.h"
#include "movie.h"
#include "stream.h"
#include "subchunk.h"
#include "util.h"

namespace ProjectorRays {

/* Movie */

void Movie::read(ReadStream *s) {
    stream = s;
    stream->endianness = kBigEndian; // we set this properly when we create the RIFX chunk
    readMemoryMap();
    readKeyTable();
    readConfig();
    readCasts();
}

void Movie::readMemoryMap() {
    // at the beginning of the file, we need to break some of the typical rules. We don't know names, lengths and offsets yet.

    // Meta
    std::shared_ptr<MetaChunk> meta = std::static_pointer_cast<MetaChunk>(readChunk(FOURCC('R', 'I', 'F', 'X')));

    if (meta->codec != FOURCC('M', 'V', '9', '3')) {
        throw std::runtime_error("Codec unsupported: " + fourCCToString(meta->codec));
    }

    // Initial map
    std::shared_ptr<InitialMapChunk> imap = std::static_pointer_cast<InitialMapChunk>(readChunk(FOURCC('i', 'm', 'a', 'p')));

    // Memory map
    stream->seek(imap->memoryMapOffset);
    std::shared_ptr<MemoryMapChunk> mmap = std::static_pointer_cast<MemoryMapChunk>(readChunk(FOURCC('m', 'm', 'a', 'p')));

    for (uint32_t i = 0; i < mmap->mapArray.size(); i++) {
        auto mapEntry = mmap->mapArray[i];

        if (mapEntry.fourCC == FOURCC('f', 'r', 'e', 'e') || mapEntry.fourCC == FOURCC('j', 'u', 'n', 'k'))
            continue;
        
        ChunkInfo info;
        info.id = i;
        info.fourCC = mapEntry.fourCC;
        info.len = mapEntry.len;
        info.uncompressedLen = mapEntry.len;
        info.offset = mapEntry.offset;
        info.compressionType = 0;
        chunkInfo[i] = info;

        chunkIDsByFourCC[mapEntry.fourCC].push_back(i);
    }
}

bool Movie::readKeyTable() {
    auto info = getFirstChunkInfo(FOURCC('K', 'E', 'Y', '*'));
    if (info) {
        keyTable = std::static_pointer_cast<KeyTableChunk>(getChunk(info->fourCC, info->id));
        return true;
    }

    std::cout << "No key chunk!\n";
    return false;
}

bool Movie::readConfig() {
    auto info = getFirstChunkInfo(FOURCC('V', 'W', 'C', 'F'));
    if (!info)
        info = getFirstChunkInfo(FOURCC('D', 'R', 'C', 'F'));

    if (info) {
        config = std::static_pointer_cast<ConfigChunk>(getChunk(info->fourCC, info->id));
        version = humanVersion(config->directorVersion);
        std::cout << "Director version: " + std::to_string(version) + "\n";

        return true;
    }

    std::cout << "No config chunk!\n";
    return false;
}

bool Movie::readCasts() {
    if (version >= 500) {
        auto info = getFirstChunkInfo(FOURCC('M', 'C', 's', 'L'));
        if (info) {
            auto castList = std::static_pointer_cast<CastListChunk>(getChunk(info->fourCC, info->id));
            for (const auto &castEntry : castList->entries) {
                std::cout << "Cast: " + castEntry.name + "\n";
                int32_t sectionID = -1;
                for (const auto &keyEntry : keyTable->entries) {
                    if (keyEntry.castID == castEntry.id && keyEntry.fourCC == FOURCC('C', 'A', 'S', '*')) {
                        sectionID = keyEntry.sectionID;
                        break;
                    }
                }
                if (sectionID > 0) {
                    auto cast = std::static_pointer_cast<CastChunk>(getChunk(FOURCC('C', 'A', 'S', '*'), sectionID));
                    cast->populate(castEntry.name, castEntry.id, castEntry.minMember);
                    casts.push_back(std::move(cast));
                }
            }

            return true;
        }

        std::cout << "No cast list!\n";
        return false;
    } else {
        auto info = getFirstChunkInfo(FOURCC('C', 'A', 'S', '*'));
        if (info) {
            auto cast = std::static_pointer_cast<CastChunk>(getChunk(info->fourCC, info->id));
            cast->populate("Internal", 1024, config->minMember);
            casts.push_back(std::move(cast));

            return true;
        }

        std::cout << "No cast!\n";
        return false;
    }

    return false;
}

const ChunkInfo *Movie::getFirstChunkInfo(uint32_t fourCC) {
    auto &chunkIDs = chunkIDsByFourCC[fourCC];
    if (chunkIDs.size() > 0) {
        return &chunkInfo[chunkIDs[0]];
    }
    return nullptr;
}

std::shared_ptr<Chunk> Movie::getChunk(uint32_t fourCC, int32_t id) {
    if (deserializedChunks.find(id) != deserializedChunks.end())
        return deserializedChunks[id];

    if (chunkInfo.find(id) == chunkInfo.end())
        throw std::runtime_error("Could not find chunk " + std::to_string(id));

    auto &info = chunkInfo[id];
    if (fourCC != info.fourCC) {
        throw std::runtime_error(
            "Expected chunk " + std::to_string(id) + " to be '" + fourCCToString(fourCC)
            + "', but is actually '" + fourCCToString(info.fourCC) + "'"
        );
    }

    stream->seek(info.offset);
    std::shared_ptr<Chunk> chunk = readChunk(fourCC, info.len);

    // don't cache the deserialized map chunks
    // we'll just generate a new one if we need to save
    if (fourCC != FOURCC('R', 'I', 'F', 'X') && fourCC != FOURCC('i', 'm', 'a', 'p') && fourCC != FOURCC('m', 'm', 'a', 'p')) {
        deserializedChunks[id] = chunk;
    }

    return chunk;
}

std::shared_ptr<Chunk> Movie::readChunk(uint32_t fourCC, uint32_t len) {
    auto offset = stream->pos();

    // check if this is the chunk we are expecting
    auto validFourCC = stream->readUint32();
    if (fourCC == FOURCC('R', 'I', 'F', 'X')) {
        //if (validName.substring(0, 2) == "MZ") {
            // handle Projector HERE
        //}
        if (validFourCC == FOURCC('X', 'F', 'I', 'R')) {
            stream->endianness = kLittleEndian;
            validFourCC = FOURCC('R', 'I', 'F', 'X');
        }
    }
    // check if it has the length the mmap table specifies
    auto validLen = stream->readUint32();

    // use the valid length if mmap hasn't been read yet
    if (len == UINT32_MAX) {
        len = validLen;
    }

    // validate chunk
    if (fourCC != validFourCC || len != validLen) {
        throw std::runtime_error(
            "At offset " + std::to_string(offset)
            + " expected '" + fourCCToString(fourCC) + "' chunk with length " + std::to_string(len)
            + ", but got '" + fourCCToString(validFourCC) + "' chunk with length " + std::to_string(validLen)
        );
    } else {
        std::cout << "At offset " + std::to_string(offset) + " reading chunk '" + fourCCToString(fourCC) + "' with length " + std::to_string(len) + "\n";
    }

    if (fourCC == FOURCC('R', 'I', 'F', 'X')) {
        // we're going to pretend RIFX is only 12 bytes long
        // this is because offsets are relative to the beginning of the file
        // whereas everywhere else they're relative to chunk start
        len = 4;
    }

    // copy the contents of the chunk to a new DataStream (minus name/length as that's not what offsets are usually relative to)
    auto chunkStream = stream->readBytes(len);

    std::shared_ptr<Chunk> res;
    switch (fourCC) {
    case FOURCC('R', 'I', 'F', 'X'):
        res = std::make_shared<MetaChunk>(this);
        break;
    case FOURCC('i', 'm', 'a', 'p'):
        res = std::make_shared<InitialMapChunk>(this);
        break;
    case FOURCC('m', 'm', 'a', 'p'):
        res = std::make_shared<MemoryMapChunk>(this);
        break;
    case FOURCC('C', 'A', 'S', '*'):
        res = std::make_shared<CastChunk>(this);
        break;
    case FOURCC('C', 'A', 'S', 't'):
        res = std::make_shared<CastMemberChunk>(this);
        break;
    case FOURCC('K', 'E', 'Y', '*'):
        res = std::make_shared<KeyTableChunk>(this);
        break;
    case FOURCC('L', 'c', 't', 'X'):
        capitalX = true;
        // fall through
    case FOURCC('L', 'c', 't', 'x'):
        res = std::make_shared<ScriptContextChunk>(this);
        break;
    case FOURCC('L', 'n', 'a', 'm'):
        res = std::make_shared<ScriptNamesChunk>(this);
        break;
    case FOURCC('L', 's', 'c', 'r'):
        res = std::make_shared<ScriptChunk>(this);
        break;
    case FOURCC('V', 'W', 'C', 'F'):
    case FOURCC('D', 'R', 'C', 'F'):
        res = std::make_shared<ConfigChunk>(this);
        break;
    case FOURCC('M', 'C', 's', 'L'):
        res = std::make_shared<CastListChunk>(this);
        break;
    default:
        res = std::make_shared<Chunk>(this);
    }

    res->read(*chunkStream);

    return res;
}

}
